internal/controllers/topology/cluster/desired_state_test.go

```diff
@@ -1603,6 +1606,105 @@ func TestComputeMachineDeployment(t *testing.T) {
+				Template: clusterv1.MachineTemplateSpec{
+					Spec: clusterv1.MachineSpec{
+						Version: pointer.String("v1.21.2"),
+						Bootstrap: clusterv1.Bootstrap{
+							ConfigRef: contract.ObjToRef(workerBootstrapTemplate),
+						},
+						InfrastructureRef: *contract.ObjToRef(workerInfrastructureMachineTemplate),
+					},
+				},
+			},
+		}
+		mdTopology = clusterv1.MachineDeploymentTopology{
+			Metadata: clusterv1.ObjectMeta{
+				Labels: map[string]string{"foo": "baz"},
+			},
+			Class: "linux-worker",
+			Name:  "big-pool-of-machines",
+		}
+		t.Run("Should set the finalizer on a new MD", func(t *testing.T) {
+			s.Current.MachineDeployments = map[string]*scope.MachineDeploymentState{
+				"big-pool-of-machines": {
+					Object:                        currentMd,
+					BootstrapTemplate:             workerBootstrapTemplate,
+					InfrastructureMachineTemplate: workerInfrastructureMachineTemplate,
+				},
+			}
+			actual, err := computeMachineDeployment(ctx, s, nil, mdTopology)
+			g.Expect(err).ToNot(HaveOccurred())
+			actualMd := actual.Object
+			g.Expect(controllerutil.ContainsFinalizer(actualMd, clusterv1.MachineDeploymentTopologyFinalizer)).To(BeTrue())
+		},
+		)
+		t.Run("Should retain finalizer on an existing MD if not set", func(t *testing.T) {
+			s.Current.MachineDeployments = map[string]*scope.MachineDeploymentState{}
+			actual, err := computeMachineDeployment(ctx, s, nil, mdTopology)
+			g.Expect(err).ToNot(HaveOccurred())
+			actualMd := actual.Object
+			g.Expect(controllerutil.ContainsFinalizer(actualMd, clusterv1.MachineDeploymentTopologyFinalizer)).To(BeTrue())
+		},
+		)
+		t.Run("Should preserve finalizer on an existing MD if set", func(t *testing.T) {
+			s.Current.MachineDeployments = map[string]*scope.MachineDeploymentState{
+				"big-pool-of-machines": {
+					Object:                        currentMd,
+					BootstrapTemplate:             workerBootstrapTemplate,
+					InfrastructureMachineTemplate: workerInfrastructureMachineTemplate,
+				},
+			}
+			currentMd.Finalizers = []string{clusterv1.MachineDeploymentTopologyFinalizer}
+			actual, err := computeMachineDeployment(ctx, s, nil, mdTopology)
+			g.Expect(err).ToNot(HaveOccurred())
+			actualMd := actual.Object
+			g.Expect(controllerutil.ContainsFinalizer(actualMd, clusterv1.MachineDeploymentTopologyFinalizer)).To(BeTrue())
+		},
+		)
+		t.Run("Should preserve finalizer on an existing MD if finalizer is already set and MD has DeletionTimestamp", func(t *testing.T) {
+			s.Current.MachineDeployments = map[string]*scope.MachineDeploymentState{
+				"big-pool-of-machines": {
+					Object:                        currentMd,
+					BootstrapTemplate:             workerBootstrapTemplate,
+					InfrastructureMachineTemplate: workerInfrastructureMachineTemplate,
+				},
+			}
+			currentMd.Finalizers = []string{clusterv1.MachineDeploymentTopologyFinalizer}
+			s.Current.MachineDeployments["big-pool-of-machines"].Object.DeletionTimestamp = &deletionTimestamp
```

Comment:
Isn't this also currentMD? If it is let's set it the same way to not confuse folks like me :) (same below)